File: backend/src/models/Tag.js
const db = require('../config/database');

class Tag {
  static async create(tagData) {
    const { name, name_th, category = 'genre' } = tagData;
    const query = `
      INSERT INTO tags (name, name_th, category)
      VALUES ($1, $2, $3)
      RETURNING *
    `;
    const result = await db.query(query, [name, name_th, category]);
    return result.rows[0];
  }

  static async findById(id) {
    const query = 'SELECT * FROM tags WHERE id = $1';
    const result = await db.query(query, [id]);
    return result.rows[0];
  }

  static async findByName(name) {
    const query = 'SELECT * FROM tags WHERE name = $1 OR name_th = $1';
    const result = await db.query(query, [name]);
    return result.rows[0];
  }

  static async update(id, tagData) {
    const { name, name_th, category } = tagData;
    const query = `
      UPDATE tags 
      SET name = COALESCE($1, name),
          name_th = COALESCE($2, name_th),
          category = COALESCE($3, category)
      WHERE id = $4
      RETURNING *
    `;
    const result = await db.query(query, [name, name_th, category, id]);
    return result.rows[0];
  }

  static async delete(id) {
    const query = 'DELETE FROM tags WHERE id = $1 RETURNING *';
    const result = await db.query(query, [id]);
    return result.rows[0];
  }

  static async getAll(category = null) {
    let query = 'SELECT * FROM tags';
    const values = [];
    
    if (category) {
      query += ' WHERE category = $1';
      values.push(category);
    }
    
    query += ' ORDER BY category, name';
    const result = await db.query(query, values);
    return result.rows;
  }

  static async search(searchQuery, limit = 10) {
    const query = `
      SELECT * FROM tags
      WHERE name ILIKE $1 OR name_th ILIKE $1
      ORDER BY name
      LIMIT $2
    `;
    const result = await db.query(query, [`%${searchQuery}%`, limit]);
    return result.rows;
  }

  static async getPopular(limit = 20) {
    const query = `
      SELECT t.*, COUNT(bt.book_id) as book_count
      FROM tags t
      LEFT JOIN book_tags bt ON t.id = bt.tag_id
      GROUP BY t.id
      ORDER BY book_count DESC
      LIMIT $1
    `;
    const result = await db.query(query, [limit]);
    return result.rows;
  }

  static async getByCategory(category) {
    const query = 'SELECT * FROM tags WHERE category = $1 ORDER BY name';
    const result = await db.query(query, [category]);
    return result.rows;
  }
}

module.exports = Tag;

File: backend/app/models/tag.py
from typing import Any, Dict, List, Optional
from uuid import UUID

from app.config.database import database


class Tag:
    @staticmethod
    async def create(tag_data: Dict[str, Any]) -> Optional[Dict[str, Any]]:
        query = """
            INSERT INTO tags (name, name_th, category)
            VALUES (:name, :name_th, :category)
            RETURNING *
        """
        values = {
            "name": tag_data.get("name"),
            "name_th": tag_data.get("name_th"),
            "category": tag_data.get("category", "genre"),
        }
        result = await database.fetch_one(query=query, values=values)
        return dict(result) if result else None

    @staticmethod
    async def find_by_id(tag_id: UUID) -> Optional[Dict[str, Any]]:
        query = "SELECT * FROM tags WHERE id = :id"
        result = await database.fetch_one(query=query, values={"id": str(tag_id)})
        return dict(result) if result else None

    @staticmethod
    async def find_by_name(name: str) -> Optional[Dict[str, Any]]:
        query = "SELECT * FROM tags WHERE name = :name OR name_th = :name"
        result = await database.fetch_one(query=query, values={"name": name})
        return dict(result) if result else None

    @staticmethod
    async def update(tag_id: UUID, tag_data: Dict[str, Any]) -> Optional[Dict[str, Any]]:
        query = """
            UPDATE tags 
            SET name = COALESCE(:name, name),
                name_th = COALESCE(:name_th, name_th),
                category = COALESCE(:category, category)
            WHERE id = :id
            RETURNING *
        """
        values = {
            "id": str(tag_id),
            "name": tag_data.get("name"),
            "name_th": tag_data.get("name_th"),
            "category": tag_data.get("category"),
        }
        result = await database.fetch_one(query=query, values=values)
        return dict(result) if result else None

    @staticmethod
    async def delete(tag_id: UUID) -> Optional[Dict[str, Any]]:
        query = "DELETE FROM tags WHERE id = :id RETURNING *"
        result = await database.fetch_one(query=query, values={"id": str(tag_id)})
        return dict(result) if result else None

    @staticmethod
    async def get_all(category: Optional[str] = None) -> List[Dict[str, Any]]:
        if category:
            query = "SELECT * FROM tags WHERE category = :category ORDER BY category, name"
            results = await database.fetch_all(query=query, values={"category": category})
        else:
            query = "SELECT * FROM tags ORDER BY category, name"
            results = await database.fetch_all(query=query)
        return [dict(r) for r in results]

    @staticmethod
    async def search(search_query: str, limit: int = 10) -> List[Dict[str, Any]]:
        query = """
            SELECT * FROM tags
            WHERE name ILIKE :search_query OR name_th ILIKE :search_query
            ORDER BY name
            LIMIT :limit
        """
        results = await database.fetch_all(
            query=query, 
            values={"search_query": f"%{search_query}%", "limit": limit}
        )
        return [dict(r) for r in results]

    @staticmethod
    async def get_popular(limit: int = 20) -> List[Dict[str, Any]]:
        query = """
            SELECT t.*, COUNT(bt.book_id) as book_count
            FROM tags t
            LEFT JOIN book_tags bt ON t.id = bt.tag_id
            GROUP BY t.id
            ORDER BY book_count DESC
            LIMIT :limit
        """
        results = await database.fetch_all(query=query, values={"limit": limit})
        return [dict(r) for r in results]

    @staticmethod
    async def get_by_category(category: str) -> List[Dict[str, Any]]:
        query = "SELECT * FROM tags WHERE category = :category ORDER BY name"
        results = await database.fetch_all(query=query, values={"category": category})
        return [dict(r) for r in results]
